File: code/algorithms/strings/LongestCommonSubstring.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <set>
#include <tuple>

class SuffixArray {
private:
    std::vector<int> sa;
    std::vector<int> lcp;
    std::vector<int> text;
    int n;

public:
    SuffixArray(const std::vector<int>& text) : text(text), n(text.size()) {}

    std::vector<int> get_sa() {
        construct_suffix_array();
        return sa;
    }

    std::vector<int> get_lcp_array() {
        build_lcp_array();
        return lcp;
    }

private:
    void construct_suffix_array() {
        std::vector<std::tuple<int, int, int>> rank_tuples(n);

        for (int i = 0; i < n; i++) {
            rank_tuples[i] = std::make_tuple(text[i], i, 0);
        }

        std::sort(rank_tuples.begin(), rank_tuples.end());

        std::vector<int> ranks(n);
        int new_rank = 0;

        for (int i = 1; i < n; i++) {
            if (std::get<0>(rank_tuples[i]) != std::get<0>(rank_tuples[i - 1])) {
                new_rank++;
            }
            ranks[std::get<1>(rank_tuples[i])] = new_rank;
        }

        for (int k = 1; k < n * 2; k *= 2) {
            std::vector<std::tuple<int, int, int>> new_rank_tuples(n);

            for (int i = 0; i < n; i++) {
                int second_half = (i + k < n) ? ranks[i + k] : -1;
                new_rank_tuples[i] = std::make_tuple(ranks[i], second_half, i);
            }

            std::sort(new_rank_tuples.begin(), new_rank_tuples.end());

            new_rank = 0;
            for (int i = 1; i < n; i++) {
                if (std::get<0>(new_rank_tuples[i]) != std::get<0>(new_rank_tuples[i - 1]) ||
                    std::get<1>(new_rank_tuples[i]) != std::get<1>(new_rank_tuples[i - 1])) {
                    new_rank++;
                }
                ranks[std::get<2>(new_rank_tuples[i])] = new_rank;
            }
        }

        sa.resize(n);
        for (int i = 0; i < n; i++) {
            sa[i] = std::get<2>(rank_tuples[i]);
        }
    }

    void build_lcp_array() {
        std::vector<int> inv_sa(n);
        for (int i = 0; i < n; i++) {
            inv_sa[sa[i]] = i;
        }

        lcp.resize(n);
        int len_lcp = 0;
        for (int i = 0; i < n; i++) {
            if (inv_sa[i] > 0) {
                int k = sa[inv_sa[i] - 1];
                while (i + len_lcp < n && k + len_lcp < n && text[i + len_lcp] == text[k + len_lcp]) {
                    len_lcp++;
                }
                lcp[inv_sa[i] - 1] = len_lcp;
                if (len_lcp > 0) {
                    len_lcp--;
                }
            }
        }
    }
};

class LongestCommonSubstring {
private:
    std::vector<std::string> strings;
    int k;
    int num_sentinels;
    int text_length;
    int shift;
    int lowest_ascii_value;
    int highest_ascii_value;
    std::vector<int> imap;
    std::vector<int> text;
    std::vector<int> sa;
    std::vector<int> lcp;
    std::set<std::string> lcss;

public:
    LongestCommonSubstring(const std::vector<std::string>& strings) : strings(strings), k(2) {
        num_sentinels = strings.size();
        compute_text_length();
        build_reverse_color_mapping();
        compute_shift();
        build_text();
    }

    void main() {
        solve();
    }

private:
    void compute_text_length() {
        text_length = 0;
        for (const std::string& s : strings) {
            text_length += s.length();
        }
        text_length += num_sentinels;
    }

    void build_reverse_color_mapping() {
        imap.clear();
        lowest_ascii_value = INT_MAX;
        highest_ascii_value = INT_MIN;

        for (int i = 0; i < strings.size(); i++) {
            for (char c : strings[i]) {
                imap.push_back(i);
                lowest_ascii_value = std::min(lowest_ascii_value, (int)c);
                highest_ascii_value = std::max(highest_ascii_value, (int)c);
            }
        }

        for (int i = 0; i < num_sentinels; i++) {
            imap.push_back(i);
        }
    }

    void compute_shift() {
        shift = num_sentinels - lowest_ascii_value;
    }

    void build_text() {
        text.clear();
        text.resize(text_length);
        int sentinel = 0;
        int k = 0;
        for (int i = 0; i < strings.size(); i++) {
            const std::string& s = strings[i];
            for (char c : s) {
                text[k++] = c + shift;
            }
            text[k++] = sentinel++;
        }
    }

    bool enough_unique_colors_in_window(int lo, int hi) {
        std::set<int> colors;
        for (int i = lo; i <= hi; i++) {
            colors.insert(imap[sa[i]]);
        }
        return colors.size() == k;
    }

    std::string retrieve_string(int i, int length) {
        std::string result;
        for (int j = 0; j < length; j++) {
            result.push_back(text[i + j] - shift);
        }
        return result;
    }

    void add_lcs(int lo, int hi, int window_lcp) {
        if (hi - lo + 1 < k || window_lcp == 0 || !enough_unique_colors_in_window(lo, hi)) {
            return;
        }

        if (window_lcp > lcss.size()) {
            lcss.clear();
        }

        if (window_lcp == lcss.size()) {
            lcss.insert(retrieve_string(sa[lo], window_lcp));
        }
    }

    void solve() {
        SuffixArray suffix_array(text);
        sa = suffix_array.get_sa();
        lcp = suffix_array.get_lcp_array();

        int lo = num_sentinels;
        int hi = num_sentinels;

        while (true) {
            bool shrink_window = (hi == text_length - 1) ? true : enough_unique_colors_in_window(lo, hi);

            if (shrink_window) {
                lo++;
            } else {
                hi++;
            }

            if (lo == text_length - 1) {
                break;
            }

            if (lo == hi) {
                continue;
            }

            int window_lcp = *std::min_element(lcp.begin() + lo, lcp.begin() + hi);
            add_lcs(lo, hi, window_lcp);
        }

        std::cout << "Longest common substrings:" << std::endl;
        for (const std::string& lcs : lcss) {
            std::cout << lcs << std::endl;
        }
    }
};

int main() {
    std::vector<std::string> strings = {"abcde", "habcab", "ghabcdf"};
    LongestCommonSubstring solver(strings);
    solver.main();
    return 0;
}

File: code/algorithms/strings/LongestCommonSubstring.py


import bisect
from collections import deque

class LongestCommonSubstring:

    def __init__(self, strings):
        self.strings = strings
        self.k = 2
        self.num_sentinels = len(strings)
        self.text_length = sum(len(s) for s in strings) + self.num_sentinels
        self.shift = 0
        self.lowest_ascii_value = float('inf')
        self.highest_ascii_value = float('-inf')
        self.imap = []
        self.text = []
        self.sa = []
        self.lcp = []
        self.lcss = set()

    def main(self):
        self.init()
        self.solve()

    def init(self):
        self.compute_text_length()
        self.build_reverse_color_mapping()
        self.compute_shift()
        self.build_text()

    def compute_text_length(self):
        self.text_length = sum(len(s) for s in self.strings) + self.num_sentinels

    def build_reverse_color_mapping(self):
        self.imap = []
        for i, s in enumerate(self.strings):
            for _ in range(len(s)):
                self.imap.append(i)
                self.lowest_ascii_value = min(self.lowest_ascii_value, ord(s[_]))
                self.highest_ascii_value = max(self.highest_ascii_value, ord(s[_]))
        for _ in range(self.num_sentinels):
            self.imap.append(_)

    def compute_shift(self):
        self.shift = self.num_sentinels - self.lowest_ascii_value

    def build_text(self):
        self.text = [0] * self.text_length
        sentinel = 0
        k = 0
        for i in range(len(self.strings)):
            s = self.strings[i]
            for j in range(len(s)):
                self.text[k] = ord(s[j]) + self.shift
                k += 1
            self.text[k] = sentinel
            k += 1
            sentinel += 1

    def enough_unique_colors_in_window(self, lo, hi):
        colors = set()
        for i in range(lo, hi + 1):
            colors.add(self.imap[self.sa[i]])
        return len(colors) == self.k

    def retrieve_string(self, i, length):
        return ''.join(chr(self.text[i + j] - self.shift) for j in range(length))

    def add_lcs(self, lo, hi, window_lcp):
        if hi - lo + 1 < self.k:
            return
        if window_lcp == 0:
            return
        if not self.enough_unique_colors_in_window(lo, hi):
            return
        if window_lcp > len(self.lcss):
            self.lcss.clear()
        if window_lcp == len(self.lcss):
            self.lcss.add(self.retrieve_string(self.sa[lo], window_lcp))

    def solve(self):
        suffix_array = SuffixArray(self.text)
        self.sa = suffix_array.get_sa()
        self.lcp = suffix_array.get_lcp_array()
        tree = CompactMinSegmentTree(self.lcp)

        lo = self.num_sentinels
        hi = self.num_sentinels

        while True:
            shrink_window = True if hi == self.text_length - 1 else self.enough_unique_colors_in_window(lo, hi)
            if shrink_window:
                lo += 1
            else:
                hi += 1
            if lo == self.text_length - 1:
                break
            if lo == hi:
                continue
            window_lcp = tree.query(lo + 1, hi + 1)
            self.add_lcs(lo, hi, window_lcp)

class SuffixArray:

    def __init__(self, text):
        self.text = text
        self.n = len(text)
        self.sa = []
        self.lcp = []

    def get_sa(self):
        self.construct_suffix_array()
        return self.sa

    def get_lcp_array(self):
        self.build_lcp_array()
        return self.lcp

    def construct_suffix_array(self):
        suffix_ranks = [[self.text[i], i] for i in range(self.n)]
        suffix_ranks.sort()
        ranks = [0] * self.n
        new_rank = 0

        for i in range(1, self.n):
            if suffix_ranks[i][0] != suffix_ranks[i - 1][0]:
                new_rank += 1
            ranks[suffix_ranks[i][1]] = new_rank

        for k in range(1, self.n * 2, k * 2):
            rank_tuples = []
            for i in range(self.n):
                second_half = ranks[i + k] if i + k < self.n else -1
                rank_tuples.append((ranks[i], second_half, i))
            rank_tuples.sort()
            new_rank = 0
            for i in range(1, self.n):
                if rank_tuples[i][0] != rank_tuples[i - 1][0] or rank_tuples[i][1] != rank_tuples[i - 1][1]:
                    new_rank += 1
                ranks[rank_tuples[i][2]] = new_rank

        self.sa = [x[2] for x in rank_tuples]

    def build_lcp_array(self):
        inv_sa = [0] * self.n
        for i in range(self.n):
            inv_sa[self.sa[i]] = i

        self.lcp = [0] * self.n
        len_lcp = 0
        for i in range(self.n):
            if inv_sa[i] > 0:
                k = self.sa[inv_sa[i] - 1]
                while i + len_lcp < self.n and k + len_lcp < self.n and self.text[i + len_lcp] == self.text[k + len_lcp]:
                    len_lcp += 1
                self.lcp[inv_sa[i] - 1] = len_lcp
                if len_lcp > 0:
                    len_lcp -= 1

class CompactMinSegmentTree:

    def __init__(self, values):
        self.n = len(values)
        self.tree = [float('inf')] * (2 * self.n)
        self.construct(values)

    def function(self, a, b):
        if a == float('inf'):
            return b
        elif b == float('inf'):
            return a
        return min(a, b)

    def modify(self, i, value):
        i += self.n
        self.tree[i] = value
        while i > 1:
            i >>= 1
            self.tree[i] = self.function(self.tree[i << 1], self.tree[i << 1 | 1])

    def query(self, l, r):
        res = float('inf')
        l += self.n
        r += self.n
        while l < r:
            if l & 1:
                res = self.function(res, self.tree[l])
                l += 1
            if r & 1:
                r -= 1
                res = self.function(res, self.tree[r])
            l >>= 1
            r >>= 1
        return res

strings = ["abcde", "habcab", "ghabcdf"]
solver = LongestCommonSubstring(strings)
solver.main()
print("Longest common substrings:", solver.lcss)
